src/libp2p_crypto.erl

```diff
@@ -370,10 +504,325 @@ base58check_decode(B58) ->
+    HashType :: atom(). % FIXME Dialyzer doesn't see multihash:hash()
+make_multisig_pubkey(M, N, PubKeys, HashType) ->
+    case lists:filter(fun pubkey_is_multisig/1, PubKeys) of
+        [_|_]=PKs ->
+            {error, {contains_multisig_keys, PKs}};
+        [] ->
+            PubKeysBins = lists:map(fun pubkey_to_bin/1, PubKeys),
+            PubKeysBin = iolist_to_binary(PubKeysBins),
+            {ok, {multisig, M, N, multihash:hash(PubKeysBin, HashType)}}
+    end.
+
+%% @doc A multisig-signature is a concatanation of a list of N
+%% individual-pubkeys and a list of M-N triples of individual-signatures
+%% prefixed with an index (of corresponsing individual-pubkey in the
+%% multisig-pubkey) and the length of the individual-signature:
+%%
+%% <<PubKeys/binary, Triples/binary>>
+%% where
+%% Pubkeys = <<PK0/binary, ..., PKN/binary>>
+%% Triples =
+%%   <<
+%%     0/integer, Len0/integer Sig0:Len0/binary,
+%%     ...,
+%%     N/integer, LenN/integer, SigN/binary
+%%   >>
+%%
+%% (see code for precise integer sizes)
+%%
+%% PubKeys MUST be in the same order as when each triple was constructed.
+%% Signature triples MAY be in any order.
+%% @end
+-spec make_multisig_signature(
+    binary(),
+    pubkey_multi(),
+    [pubkey_single()],
+    [{non_neg_integer(), binary()}]
+) -> {ok, binary()} | {error, Error} when
+    Error ::
+        insufficient_signatures
+        | insufficient_keys
+        | too_many_keys
+        | bad_key_digest
+        | {invalid_signatures, [binary()]}.
+make_multisig_signature(_, {multisig, M, _, _}, _, S) when M > length(S) ->
+    {error, insufficient_signatures};
+make_multisig_signature(_, {multisig, _, N, _}, K, _) when N > length(K) ->
+    {error, insufficient_keys};
+make_multisig_signature(_, {multisig, _, N, _}, K, _) when N < length(K) ->
+    {error, too_many_keys};
+make_multisig_signature(Msg, {multisig, _, _, KeysDigest}, Keys, ISigs) ->
+    {ok, _, HashType, _} = multihash:decode(KeysDigest),
+    KeysBin = iolist_to_binary(lists:map(fun pubkey_to_bin/1, Keys)),
+    case multihash:hash(KeysBin, HashType) of
+        <<KeysDigest/binary>> ->
+            KeySigs = [{lists:nth(I + 1, Keys), S} || {I, S} <- ISigs],
+            case [S || {K, S} <- KeySigs, not verify(Msg, S, K)] of
+                [] ->
+                    ISigsBins = lists:map(fun isig_to_bin/1, ISigs),
+                    {ok, iolist_to_binary([KeysBin, ISigsBins])};
+                [_|_]=Sigs ->
+                    {error, {invalid_signatures, Sigs}}
+            end;
+        <<_/binary>> ->
+            {error, bad_key_digest}
+    end.
+
+isig_to_bin({I, <<Sig/binary>>}) ->
+    Len = byte_size(Sig),
+    <<
+        I:?MULTISIG_KEY_INDEX_BITS/integer-unsigned-little,
+        Len:?MULTISIG_SIG_LEN_BYTES/integer-unsigned-little,
+        Sig/binary
+    >>.
+
 -ifdef(TEST).
 
 -include_lib("eunit/include/eunit.hrl").
 
+make_multisig_test_cases(M, N, HashType, KeyType) ->
```

Comment:
All of this should really be in a common test imo. Would make it _much_ simpler to read through and update in future.

Reply:
If you mean just the volume of it - I can see where you are coming from...

These tests are all stateless, IO-less, simple things though; _and_ some internal functions are used too (`make_multisig_pubkey_`) - which puts them squarely within unit-testing domain.

There's also the question of feature-parity - I don't think you can do the equivalent test-case-generation (with title annotations and all) with CT.